overlord/devicestate/firstboot_preseed_test.go

```diff
@@ -0,0 +1,224 @@
+	c.Assert(preseedTask.Kind(), Equals, "mark-preseeded")
+	c.Check(markSeededTask.Kind(), Equals, "mark-seeded")
+
+	// mark-seeded waits for mark-preseeded
+	var waitsForPreseeded bool
+	for _, wt := range markSeededTask.WaitTasks() {
+		if wt.Kind() == "mark-preseeded" {
+			waitsForPreseeded = true
+		}
+	}
+	c.Check(waitsForPreseeded, Equals, true)
+}
+
+func checkPresseedTaskStates(c *C, st *state.State) {
+	doneTasks := map[string]bool{
+		"prerequisites":        true,
+		"prepare-snap":         true,
+		"link-snap":            true,
+		"mount-snap":           true,
+		"setup-profiles":       true,
+		"update-gadget-assets": true,
+		"copy-snap-data":       true,
+		"set-auto-aliases":     true,
+		"setup-aliases":        true,
+		"auto-connect":         true,
+	}
+	doTasks := map[string]bool{
+		"run-hook":            true,
+		"mark-seeded":         true,
+		"start-snap-services": true,
+	}
+	seenDone := make(map[string]bool)
+	for _, t := range st.Tasks() {
+		if t.Status() == state.DoneStatus {
+			seenDone[t.Kind()] = true
+		}
+		switch {
+		case doneTasks[t.Kind()]:
+			c.Check(t.Status(), Equals, state.DoneStatus, Commentf("task: %s", t.Kind()))
+		case t.Kind() == "mark-preseeded":
+			c.Check(t.Status(), Equals, state.DoingStatus, Commentf("task: %s", t.Kind()))
+		case doTasks[t.Kind()]:
+			c.Check(t.Status(), Equals, state.DoStatus, Commentf("task: %s", t.Kind()))
+		default:
+			c.Fatalf("unhandled task kind %s", t.Kind())
+		}
+	}
+	// sanity: check that doneTasks is not declaring more tasks than
+	// actually expected.
+	c.Check(doneTasks, DeepEquals, seenDone)
+}
+
+func markPreseededInWaitChain(t *state.Task) bool {
+	for _, wt := range t.WaitTasks() {
+		if wt.Kind() == "mark-preseeded" {
+			return true
+		}
+		if markPreseededInWaitChain(wt) {
+			return true
+		}
+	}
+	return false
+}
+
+func checkPreseedOrder(c *C, tsAll []*state.TaskSet, snaps ...string) {
+	matched := 0
+	markSeeded := 0
+	markPreseeded := 0
+	markPreseededWaitingForAliases := 0
+
+	for _, ts := range tsAll {
+		for _, t := range ts.Tasks() {
+			switch t.Kind() {
+			case "run-hook":
+				// ensure that hooks are run after mark-preseeded
+				c.Check(markPreseededInWaitChain(t), Equals, true)
+			case "mark-seeded":
+				// nothing waits for mark-seeded
+				c.Check(t.HaltTasks(), HasLen, 0)
+				markSeeded++
+				c.Check(markPreseededInWaitChain(t), Equals, true)
+			case "mark-preseeded":
+				for _, wt := range t.WaitTasks() {
+					if wt.Kind() == "setup-aliases" {
+						markPreseededWaitingForAliases++
+					}
+				}
+				markPreseeded++
+			}
+		}
+	}
+
+	c.Check(markSeeded, Equals, 1)
+	c.Check(markPreseeded, Equals, 1)
+	c.Check(markPreseededWaitingForAliases, Equals, len(snaps))
+
+	// check that prerequisites tasks for all snaps are present and
+	// are chained properly.
+	var prevTask *state.Task
+	for i, ts := range tsAll {
+		task0 := ts.Tasks()[0]
+		waitTasks := task0.WaitTasks()
+		if task0.Kind() != "prerequisites" {
```

Comment:
when would this be the case that prerequisites is not the first task?

Reply:
Added a comment explaining, plus extra check for the "else" case.